File: src/pi_wave/plan.py
"""Wave plan loading and validation.

A plan is a JSON file:
{
  "name": "my-run",
  "cwd": "/abs/path/to/project",          // optional, default: process cwd
  "on_failure": "stop",                   // "stop" (default) | "continue"
  "waves": [
    [ {assignment}, {assignment} ],       // one wave = parallel batch
    [ {assignment} ]
  ]
}

Assignment:
{
  "name": "wave-1-research",              // unique across the whole plan
  "prompt": "self-contained instructions",
  "model": "openai-codex/gpt-5.6-luna",   // provider-qualified id (pi kind)
                                          // omp kind: bare id, e.g. "claude-sonnet-5"
  "provider": "openai-codex",             // optional (redundant if model is qualified)
  "kind": "pi",                           // "pi" (default) | "omp" (Sonnet via OMP;
                                          // requires display=herdr)
  "thinking": "off|minimal|low|medium|high|xhigh|max",   // default "high"
  "files": ["src/a.py"],                  // exact files this agent may touch ([] = read-only)
  "done_when": "definition of done text", // optional, goes into the prompt
  "review_cmd": "pytest -q",              // optional; exit 0 = pass, else fix round
  "max_fix_rounds": 2,
  "timeout": 600,                         // seconds per prompt round
  "needs_results": ["wave-1-research"],   // inject prior agents' final texts
  "load_extensions": false,               // keep false: blocks nested dispatch_wave
  "display": "auto",                      // "auto" (default) — herdr panes when run
}                                         // inside Herdr, headless otherwise;
                                          // "herdr"/"headless" force one mode
// "mcp_config": "/path/to/mcp.json"     // optional; gives the agent MCP tools
//                                       // (e.g. wigolo research) via pi-mcp-adapter

Plan-level:
{
  "orchestrator": true,   // or {"model": "...", "thinking": "high",
}                         //     "synthesis_model": "...", "synthesis_thinking": "high"}
                          // spawns the delegate-wave orchestrator pane (default
                          // openai-codex/gpt-5.6-sol at high, wider pane). It
                          // reviews results without a review_cmd and writes the
                          // final synthesis. Set synthesis_model/synthesis_thinking
                          // to hand ONLY the synthesis to a dedicated
                          // `synthesizer` pane. Herdr display only.

Enforced delegate-wave rules:
- two assignments in the same wave must not touch the same file;
- names are unique plan-wide.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from pathlib import Path

THINKING_LEVELS = {"off", "minimal", "low", "medium", "high", "xhigh", "max"}
DISPLAY_MODES = {"auto", "headless", "herdr"}
AGENT_KINDS = {"pi", "omp"}
NAME_PATTERN = r"^[a-z][a-z0-9_-]{0,31}$"
ORCHESTRATOR_NAME = "orchestrator"
SYNTHESIZER_NAME = "synthesizer"


def resolve_display(display: str) -> str:
    """auto → herdr when running inside Herdr, headless otherwise."""
    if display == "auto":
        return "herdr" if os.environ.get("HERDR_ENV") == "1" else "headless"
    return display


class PlanError(ValueError):
    pass


@dataclass
class Assignment:
    name: str
    prompt: str
    model: str
    provider: str | None = None
    kind: str = "pi"
    thinking: str = "high"
    files: list[str] = field(default_factory=list)
    done_when: str = ""
    review_cmd: str = ""
    max_fix_rounds: int = 2
    timeout: float = 600.0
    needs_results: list[str] = field(default_factory=list)
    load_extensions: bool = False
    display: str = "auto"
    mcp_config: str = ""


@dataclass
class OrchestratorSpec:
    """The delegate-wave orchestrator pane: reviews results against done_when
    and writes the final synthesis. Defaults follow the skill (sol @ high).
    synthesis_model/synthesis_thinking, when set, hand ONLY the final
    synthesis to a dedicated one-shot `synthesizer` pane; the orchestrator
    pane keeps doing the reviews."""
    model: str = "openai-codex/gpt-5.6-sol"
    thinking: str = "high"
    synthesis_model: str = ""
    synthesis_thinking: str = ""


@dataclass
class Plan:
    name: str
    cwd: str
    on_failure: str
    waves: list[list[Assignment]]
    source: str = ""
    display: str = "auto"
    orchestrator: OrchestratorSpec | None = None


def _req(d: dict, key: str, ctx: str):
    if key not in d or not str(d[key]).strip():
        raise PlanError(f"{ctx}: missing required field '{key}'")
    return d[key]


def _parse_assignment(d: dict, wave_idx: int, idx: int, default_display: str = "headless") -> Assignment:
    ctx = f"wave {wave_idx + 1}, agent {idx + 1}"
    if not isinstance(d, dict):
        raise PlanError(f"{ctx}: assignment must be an object")
    import re

    name = str(_req(d, "name", ctx))
    if not re.match(NAME_PATTERN, name):
        raise PlanError(f"{ctx}: name '{name}' must match {NAME_PATTERN}")
    prompt = str(_req(d, "prompt", ctx))
    model = str(_req(d, "model", ctx))
    provider = str(d["provider"]).strip() if d.get("provider") else None
    kind = str(d.get("kind", "pi"))
    if kind not in AGENT_KINDS:
        raise PlanError(f"{ctx}: kind '{kind}' not in {sorted(AGENT_KINDS)}")
    if kind == "pi" and "/" not in model and not provider:
        raise PlanError(
            f"{ctx}: model '{model}' is not provider-qualified and no 'provider' "
            f"field given (pi requires 'provider/model', its default provider is google)"
        )
    thinking = str(d.get("thinking", "high"))
    if thinking not in THINKING_LEVELS:
        raise PlanError(f"{ctx}: thinking '{thinking}' not in {sorted(THINKING_LEVELS)}")
    files = d.get("files", [])
    if not isinstance(files, list) or not all(isinstance(f, str) for f in files):
        raise PlanError(f"{ctx}: 'files' must be a list of strings")
    timeout = float(d.get("timeout", 600))
    if timeout <= 0:
        raise PlanError(f"{ctx}: timeout must be > 0")
    max_fix = int(d.get("max_fix_rounds", 2))
    if max_fix < 0 or max_fix > 5:
        raise PlanError(f"{ctx}: max_fix_rounds must be 0..5")
    display = str(d.get("display", default_display))
    if display not in DISPLAY_MODES:
        raise PlanError(f"{ctx}: display '{display}' not in {sorted(DISPLAY_MODES)}")
    mcp_config = str(d.get("mcp_config", "")).strip()
    if mcp_config and not Path(mcp_config).expanduser().is_file():
        raise PlanError(f"{ctx}: mcp_config file not found: {mcp_config}")
    return Assignment(
        name=name,
        prompt=prompt,
        model=model,
        provider=provider,
        kind=kind,
        thinking=thinking,
        files=[os.path.normpath(f) for f in files],
        done_when=str(d.get("done_when", "")),
        review_cmd=str(d.get("review_cmd", "")),
        max_fix_rounds=max_fix,
        timeout=timeout,
        needs_results=[str(n) for n in d.get("needs_results", [])],
        load_extensions=bool(d.get("load_extensions", False)),
        display=display,
        mcp_config=mcp_config,
    )

def _parse_orchestrator(v, p: Path) -> OrchestratorSpec | None:
    """true → skill defaults; {"model": ..., "thinking": ...} → custom."""
    if not v:
        return None
    spec = OrchestratorSpec()
    if isinstance(v, dict):
        if v.get("model"):
            spec.model = str(v["model"]).strip()
        if v.get("thinking"):
            spec.thinking = str(v["thinking"]).strip()
        if v.get("synthesis_model"):
            spec.synthesis_model = str(v["synthesis_model"]).strip()
        if v.get("synthesis_thinking"):
            spec.synthesis_thinking = str(v["synthesis_thinking"]).strip()
    elif v is not True:
        raise PlanError(
            f"{p}: 'orchestrator' must be true or an object with model/thinking"
        )
    if "/" not in spec.model:
        raise PlanError(
            f"{p}: orchestrator model '{spec.model}' is not provider-qualified "
            f"(the orchestrator runs via pi, which requires 'provider/model')"
        )
    if spec.thinking not in THINKING_LEVELS:
        raise PlanError(
            f"{p}: orchestrator thinking '{spec.thinking}' not in {sorted(THINKING_LEVELS)}"
        )
    if spec.synthesis_model and "/" not in spec.synthesis_model:
        raise PlanError(
            f"{p}: synthesis_model '{spec.synthesis_model}' is not provider-qualified "
            f"(the synthesizer runs via pi, which requires 'provider/model')"
        )
    if spec.synthesis_thinking and spec.synthesis_thinking not in THINKING_LEVELS:
        raise PlanError(
            f"{p}: synthesis_thinking '{spec.synthesis_thinking}' not in {sorted(THINKING_LEVELS)}"
        )
    return spec


def load_plan(path: str | Path) -> Plan:
    p = Path(path).expanduser()
    if not p.is_file():
        raise PlanError(f"plan file not found: {p}")
    try:
        raw = json.loads(p.read_text(encoding="utf-8"))
    except json.JSONDecodeError as e:
        raise PlanError(f"{p}: invalid JSON: {e}") from e
    if not isinstance(raw, dict):
        raise PlanError(f"{p}: top level must be an object")
    waves_raw = raw.get("waves")
    if not isinstance(waves_raw, list) or not waves_raw:
        raise PlanError(f"{p}: 'waves' must be a non-empty list of waves")

    plan = Plan(
        name=str(raw.get("name", p.stem)),
        cwd=str(Path(str(raw.get("cwd", "")) or Path.cwd()).expanduser().resolve()),
        on_failure=str(raw.get("on_failure", "stop")),
        waves=[],
        source=str(p),
        display=str(raw.get("display", "auto")),
        orchestrator=_parse_orchestrator(raw.get("orchestrator"), p),
    )
    if plan.display not in DISPLAY_MODES:
        raise PlanError(f"{p}: display must be one of {sorted(DISPLAY_MODES)}")
    if plan.on_failure not in {"stop", "continue"}:
        raise PlanError(f"{p}: on_failure must be 'stop' or 'continue'")
    if not Path(plan.cwd).is_dir():
        raise PlanError(f"{p}: cwd does not exist: {plan.cwd}")

    seen_names: set[str] = set()
    all_names: list[str] = []
    for w_idx, wave_raw in enumerate(waves_raw):
        if not isinstance(wave_raw, list) or not wave_raw:
            raise PlanError(f"{p}: wave {w_idx + 1} must be a non-empty list of assignments")
        wave: list[Assignment] = []
        seen_files: dict[str, str] = {}
        for a_idx, a_raw in enumerate(wave_raw):
            a = _parse_assignment(a_raw, w_idx, a_idx, default_display=plan.display)
            if a.name in seen_names:
                raise PlanError(f"{p}: duplicate agent name '{a.name}' (names must be unique plan-wide)")
            if plan.orchestrator and a.name == ORCHESTRATOR_NAME:
                raise PlanError(
                    f"{p}: assignment name '{ORCHESTRATOR_NAME}' is reserved for the "
                    f"orchestrator pane when plan-level 'orchestrator' is enabled"
                )
            if (plan.orchestrator and plan.orchestrator.synthesis_model
                    and a.name == SYNTHESIZER_NAME):
                raise PlanError(
                    f"{p}: assignment name '{SYNTHESIZER_NAME}' is reserved for the "
                    f"synthesis pane when orchestrator synthesis_model is set"
                )
            seen_names.add(a.name)
            for f in a.files:
                if f in seen_files:
                    raise PlanError(
                        f"{p}: file conflict in wave {w_idx + 1}: '{f}' is touched by both "
                        f"'{seen_files[f]}' and '{a.name}' — move one to a later wave"
                    )
                seen_files[f] = a.name
            wave.append(a)
            all_names.append(a.name)
        plan.waves.append(wave)

    seen_before: set[str] = set()
    for w_idx, wave in enumerate(plan.waves):
        for a in wave:
            for dep in a.needs_results:
                if dep not in all_names:
                    raise PlanError(f"{p}: '{a.name}' needs results of unknown agent '{dep}'")
                if dep not in seen_before:
                    raise PlanError(
                        f"{p}: '{a.name}' needs results of '{dep}', which is not in an earlier wave"
                    )
        seen_before.update(a.name for a in wave)
    return plan


def apply_overrides(plan: Plan, model: str | None = None,
                    provider: str | None = None,
                    thinking: str | None = None) -> Plan:
    """Replace model/provider/thinking on every assignment (CLI overrides).

    Mutates and returns the same plan. Same qualification rule as
    _parse_assignment: an unqualified model needs a provider — from the
    override or from the assignment itself (an assignment-level provider
    survives a --model-only override).
    """
    if model is not None and not model.strip():
        raise PlanError("override: model must not be empty")
    if thinking is not None and thinking not in THINKING_LEVELS:
        raise PlanError(f"override: thinking '{thinking}' not in {sorted(THINKING_LEVELS)}")
    for wave in plan.waves:
        for a in wave:
            if model is not None:
                a.model = model.strip()
            if provider is not None:
                a.provider = provider.strip() or None
            if thinking is not None:
                a.thinking = thinking
            if a.kind == "pi" and "/" not in a.model and not a.provider:
                raise PlanError(
                    f"override: agent '{a.name}' would get unqualified model "
                    f"'{a.model}' and no provider (pi requires 'provider/model'; "
                    f"pass --provider too)"
                )
    return plan

File: engine/plan.ts
// Wave plan loading and validation. The full plan schema is in README.md
// ("Wave plan schema"). Field names stay snake_case so they match the plan
// JSON one to one.
//
// Enforced delegate-wave rules:
// - two assignments in the same wave must not touch the same file;
// - names are unique plan-wide.

import { readFileSync, realpathSync, statSync } from "node:fs";
import os from "node:os";
import path from "node:path";

export const THINKING_LEVELS = ["off", "minimal", "low", "medium", "high", "xhigh", "max"] as const;
export const DISPLAY_MODES = ["auto", "headless", "herdr"] as const;
export const AGENT_KINDS = ["pi", "omp"] as const;
export const NAME_PATTERN = /^[a-z][a-z0-9_-]{0,31}$/;
export const ORCHESTRATOR_NAME = "orchestrator";
export const SYNTHESIZER_NAME = "synthesizer";

export type Thinking = (typeof THINKING_LEVELS)[number];
export type Display = (typeof DISPLAY_MODES)[number];
export type AgentKind = (typeof AGENT_KINDS)[number];

export class PlanError extends Error {
  override name = "PlanError";
}

export interface Assignment {
  name: string;
  prompt: string;
  model: string;
  provider: string | null;
  kind: AgentKind;
  thinking: Thinking;
  files: string[];
  done_when: string;
  review_cmd: string;
  max_fix_rounds: number;
  timeout: number;
  needs_results: string[];
  load_extensions: boolean;
  display: Display;
  mcp_config: string;
}

/** The delegate-wave orchestrator pane: reviews results against done_when
 * and writes the final synthesis. Defaults follow the skill (gpt-5.5 @ high).
 * synthesis_model/synthesis_thinking, when set, hand ONLY the final
 * synthesis to a dedicated one-shot `synthesizer` pane; the orchestrator
 * pane keeps doing the reviews. */
export interface OrchestratorSpec {
  model: string;
  thinking: string;
  synthesis_model: string;
  synthesis_thinking: string;
}

export interface Plan {
  name: string;
  cwd: string;
  on_failure: "stop" | "continue";
  waves: Assignment[][];
  source: string;
  display: Display;
  orchestrator: OrchestratorSpec | null;
}

/** auto → herdr when running inside Herdr, headless otherwise. */
export function resolveDisplay(display: Display): "herdr" | "headless" {
  if (display === "auto") return process.env.HERDR_ENV === "1" ? "herdr" : "headless";
  return display;
}

type Raw = Record<string, unknown>;

const isObject = (v: unknown): v is Raw =>
  typeof v === "object" && v !== null && !Array.isArray(v);

const oneOf = <T extends string>(values: readonly T[], v: string): v is T =>
  (values as readonly string[]).includes(v);

const listing = (values: readonly string[]) =>
  `[${[...values].sort().map((v) => `'${v}'`).join(", ")}]`;

const expandUser = (p: string) =>
  p === "~" || p.startsWith("~/") ? path.join(os.homedir(), p.slice(1)) : p;

/** Like Python's Path.is_file/is_dir: any stat error (ENOENT, ENOTDIR,
 * ENAMETOOLONG, ...) means "no". */
const statOrNull = (p: string) => {
  try {
    return statSync(p);
  } catch {
    return null;
  }
};
const isFile = (p: string) => statOrNull(p)?.isFile() ?? false;
const isDir = (p: string) => statOrNull(p)?.isDirectory() ?? false;

/** Like Python's os.path.normpath: also drops a trailing separator, so
 * "src/" and "src" count as the same file for conflict detection. */
const normPath = (p: string) => {
  const n = path.normalize(p);
  return n.length > 1 && n.endsWith(path.sep) ? n.slice(0, -1) : n;
};

const str = (v: unknown, fallback = "") => (v === undefined || v === null ? fallback : String(v));

function req(d: Raw, key: string, ctx: string): string {
  const v = str(d[key]);
  if (!v.trim()) throw new PlanError(`${ctx}: missing required field '${key}'`);
  return v;
}

function isStringList(v: unknown): v is string[] {
  return Array.isArray(v) && v.every((x) => typeof x === "string");
}

function parseAssignment(d: unknown, waveIdx: number, idx: number, defaultDisplay: Display): Assignment {
  const ctx = `wave ${waveIdx + 1}, agent ${idx + 1}`;
  if (!isObject(d)) throw new PlanError(`${ctx}: assignment must be an object`);

  const name = req(d, "name", ctx);
  if (!NAME_PATTERN.test(name)) {
    throw new PlanError(`${ctx}: name '${name}' must match ${NAME_PATTERN.source}`);
  }
  const prompt = req(d, "prompt", ctx);
  const model = req(d, "model", ctx);
  const provider = d.provider ? String(d.provider).trim() || null : null;
  const kind = str(d.kind, "pi");
  if (!oneOf(AGENT_KINDS, kind)) {
    throw new PlanError(`${ctx}: kind '${kind}' not in ${listing(AGENT_KINDS)}`);
  }
  if (kind === "pi" && !model.includes("/") && !provider) {
    throw new PlanError(
      `${ctx}: model '${model}' is not provider-qualified and no 'provider' ` +
        `field given (pi requires 'provider/model', its default provider is google)`,
    );
  }
  const thinking = str(d.thinking, "high");
  if (!oneOf(THINKING_LEVELS, thinking)) {
    throw new PlanError(`${ctx}: thinking '${thinking}' not in ${listing(THINKING_LEVELS)}`);
  }
  const files = d.files ?? [];
  if (!isStringList(files)) throw new PlanError(`${ctx}: 'files' must be a list of strings`);
  const timeout = Number(d.timeout ?? 600);
  if (!Number.isFinite(timeout) || timeout <= 0) throw new PlanError(`${ctx}: timeout must be > 0`);
  const maxFix = Number(d.max_fix_rounds ?? 2);
  if (!Number.isInteger(maxFix) || maxFix < 0 || maxFix > 5) {
    throw new PlanError(`${ctx}: max_fix_rounds must be 0..5`);
  }
  const display = str(d.display, defaultDisplay);
  if (!oneOf(DISPLAY_MODES, display)) {
    throw new PlanError(`${ctx}: display '${display}' not in ${listing(DISPLAY_MODES)}`);
  }
  const mcpConfig = str(d.mcp_config).trim();
  if (mcpConfig && !isFile(expandUser(mcpConfig))) {
    throw new PlanError(`${ctx}: mcp_config file not found: ${mcpConfig}`);
  }
  const needs = d.needs_results ?? [];
  if (!isStringList(needs)) throw new PlanError(`${ctx}: 'needs_results' must be a list of strings`);
  return {
    name,
    prompt,
    model,
    provider,
    kind,
    thinking,
    files: files.map(normPath),
    done_when: str(d.done_when),
    review_cmd: str(d.review_cmd),
    max_fix_rounds: maxFix,
    timeout,
    needs_results: needs,
    load_extensions: Boolean(d.load_extensions),
    display,
    mcp_config: mcpConfig,
  };
}

/** true → skill defaults; {"model": ..., "thinking": ...} → custom.
 * Falsy values and an empty object disable it, as in the Python engine. */
function parseOrchestrator(v: unknown, p: string): OrchestratorSpec | null {
  if (!v || (isObject(v) && Object.keys(v).length === 0) || (Array.isArray(v) && v.length === 0)) {
    return null;
  }
  const spec: OrchestratorSpec = {
    model: "openai-codex/gpt-5.5",
    thinking: "high",
    synthesis_model: "",
    synthesis_thinking: "",
  };
  if (isObject(v)) {
    for (const key of ["model", "thinking", "synthesis_model", "synthesis_thinking"] as const) {
      if (v[key]) spec[key] = String(v[key]).trim();
    }
  } else if (v !== true) {
    throw new PlanError(`${p}: 'orchestrator' must be true or an object with model/thinking`);
  }
  if (!spec.model.includes("/")) {
    throw new PlanError(
      `${p}: orchestrator model '${spec.model}' is not provider-qualified ` +
        `(the orchestrator runs via pi, which requires 'provider/model')`,
    );
  }
  if (!oneOf(THINKING_LEVELS, spec.thinking)) {
    throw new PlanError(
      `${p}: orchestrator thinking '${spec.thinking}' not in ${listing(THINKING_LEVELS)}`,
    );
  }
  if (spec.synthesis_model && !spec.synthesis_model.includes("/")) {
    throw new PlanError(
      `${p}: synthesis_model '${spec.synthesis_model}' is not provider-qualified ` +
        `(the synthesizer runs via pi, which requires 'provider/model')`,
    );
  }
  if (spec.synthesis_thinking && !oneOf(THINKING_LEVELS, spec.synthesis_thinking)) {
    throw new PlanError(
      `${p}: synthesis_thinking '${spec.synthesis_thinking}' not in ${listing(THINKING_LEVELS)}`,
    );
  }
  return spec;
}

/** Load a plan file. A relative planPath, and a relative or omitted plan
 * `cwd`, resolve against baseDir (the caller's working directory). */
export function loadPlan(planPath: string, baseDir: string = process.cwd()): Plan {
  const p = path.resolve(baseDir, expandUser(planPath));
  if (!isFile(p)) throw new PlanError(`plan file not found: ${p}`);
  let raw: unknown;
  try {
    raw = JSON.parse(readFileSync(p, "utf8"));
  } catch (e) {
    throw new PlanError(`${p}: invalid JSON: ${(e as Error).message}`);
  }
  return parsePlan(raw, p, baseDir);
}

/** Validate an already-parsed plan object. `source` names it in errors
 * (and, without a plan `name`, names the plan); a relative or omitted plan
 * `cwd` resolves against baseDir. */
export function parsePlan(raw: unknown, source: string, baseDir: string = process.cwd()): Plan {
  const p = source;
  if (!isObject(raw)) throw new PlanError(`${p}: top level must be an object`);
  const wavesRaw = raw.waves;
  if (!Array.isArray(wavesRaw) || wavesRaw.length === 0) {
    throw new PlanError(`${p}: 'waves' must be a non-empty list of waves`);
  }

  const name = str(raw.name, path.parse(p).name);
  const cwd = path.resolve(baseDir, expandUser(str(raw.cwd) || "."));
  const onFailure = str(raw.on_failure, "stop");
  const display = str(raw.display, "auto");
  const orchestrator = parseOrchestrator(raw.orchestrator, p);
  if (!oneOf(DISPLAY_MODES, display)) {
    throw new PlanError(`${p}: display must be one of ${listing(DISPLAY_MODES)}`);
  }
  if (onFailure !== "stop" && onFailure !== "continue") {
    throw new PlanError(`${p}: on_failure must be 'stop' or 'continue'`);
  }
  if (!isDir(cwd)) throw new PlanError(`${p}: cwd does not exist: ${cwd}`);

  const plan: Plan = {
    name,
    cwd: realpathSync(cwd),
    on_failure: onFailure,
    waves: [],
    source: p,
    display,
    orchestrator,
  };

  const seenNames = new Set<string>();
  wavesRaw.forEach((waveRaw, wIdx) => {
    if (!Array.isArray(waveRaw) || waveRaw.length === 0) {
      if (isObject(waveRaw)) {
        throw new PlanError(
          `${p}: wave ${wIdx + 1} is a single assignment object, but each wave must be a list of ` +
            `assignments - write waves as [[{...}, {...}], [{...}]]`,
        );
      }
      throw new PlanError(`${p}: wave ${wIdx + 1} must be a non-empty list of assignments`);
    }
    const wave: Assignment[] = [];
    const seenFiles = new Map<string, string>();
    waveRaw.forEach((aRaw, aIdx) => {
      const a = parseAssignment(aRaw, wIdx, aIdx, display);
      if (seenNames.has(a.name)) {
        throw new PlanError(`${p}: duplicate agent name '${a.name}' (names must be unique plan-wide)`);
      }
      if (orchestrator && a.name === ORCHESTRATOR_NAME) {
        throw new PlanError(
          `${p}: assignment name '${ORCHESTRATOR_NAME}' is reserved for the ` +
            `orchestrator pane when plan-level 'orchestrator' is enabled`,
        );
      }
      if (orchestrator?.synthesis_model && a.name === SYNTHESIZER_NAME) {
        throw new PlanError(
          `${p}: assignment name '${SYNTHESIZER_NAME}' is reserved for the ` +
            `synthesis pane when orchestrator synthesis_model is set`,
        );
      }
      seenNames.add(a.name);
      for (const f of a.files) {
        const owner = seenFiles.get(f);
        if (owner !== undefined) {
          throw new PlanError(
            `${p}: file conflict in wave ${wIdx + 1}: '${f}' is touched by both ` +
              `'${owner}' and '${a.name}' - move one to a later wave`,
          );
        }
        seenFiles.set(f, a.name);
      }
      wave.push(a);
    });
    plan.waves.push(wave);
  });

  const seenBefore = new Set<string>();
  for (const wave of plan.waves) {
    for (const a of wave) {
      for (const dep of a.needs_results) {
        if (!seenNames.has(dep)) {
          throw new PlanError(`${p}: '${a.name}' needs results of unknown agent '${dep}'`);
        }
        if (!seenBefore.has(dep)) {
          throw new PlanError(`${p}: '${a.name}' needs results of '${dep}', which is not in an earlier wave`);
        }
      }
    }
    for (const a of wave) seenBefore.add(a.name);
  }
  return plan;
}

export interface Overrides {
  model?: string;
  provider?: string;
  thinking?: string;
}

/** Replace model/provider/thinking on every assignment (CLI overrides).
 *
 * Mutates and returns the same plan. Same qualification rule as
 * parseAssignment: an unqualified model needs a provider - from the
 * override or from the assignment itself (an assignment-level provider
 * survives a --model-only override). */
export function applyOverrides(plan: Plan, { model, provider, thinking }: Overrides = {}): Plan {
  if (model !== undefined && !model.trim()) throw new PlanError("override: model must not be empty");
  if (thinking !== undefined && !oneOf(THINKING_LEVELS, thinking)) {
    throw new PlanError(`override: thinking '${thinking}' not in ${listing(THINKING_LEVELS)}`);
  }
  for (const wave of plan.waves) {
    for (const a of wave) {
      if (model !== undefined) a.model = model.trim();
      if (provider !== undefined) a.provider = provider.trim() || null;
      if (thinking !== undefined) a.thinking = thinking;
      if (a.kind === "pi" && !a.model.includes("/") && !a.provider) {
        throw new PlanError(
          `override: agent '${a.name}' would get unqualified model ` +
            `'${a.model}' and no provider (pi requires 'provider/model'; ` +
            `pass --provider too)`,
        );
      }
    }
  }
  return plan;
}
